### PR TITLE
FRPAL-4732 - Adjustments for Agent Profile

### PR DESCRIPTION
**Pull request by bbeversdorf**

### CODE REVIEW

File: Float.TinCan/RemoteLRS.cs [hunk 1]
```diff
@@ -725,7 +727,7 @@ async Task<LRSResponse> SaveDocument(string resource, Dictionary<string, string>
                 document.contentType = "application/json";
             }
 
-            var contentType = MediaTypeHeaderValue.Parse(document.contentType);
+            var contentType = MediaTypeHeaderValue.TryParse(document.contentType, out var contentMediaType) ? contentMediaType : MediaTypeHeaderValue.Parse("application/json");
```
Comment (mediabounds) [inline Float.TinCan/RemoteLRS.cs:730]: @bbeversdorf Is this change necessary? The old version would have thrown an exception if the `document.contentType` was invalid (which I tend to think the implementing application should be aware of).

File: Float.TinCan/RemoteLRS.cs [hunk 2]
```diff
@@ -494,10 +494,12 @@ public async Task<AgentProfileLRSResponse> RetrieveAgentProfile(string id, Agent
...
 
         /// <inheritdoc/>
-        public Task<LRSResponse> SaveAgentProfile(AgentProfileDocument profile)
+        public async Task<LRSResponse> SaveAgentProfile(AgentProfileDocument profile)
         {
             Contract.Requires(profile != null);
-            return SaveAgentProfile(profile, RequestType.put);
+            var currentProfile = await RetrieveAgentProfile(profile.id, profile.agent);
+            profile.etag = currentProfile?.content?.etag;
+            return await SaveAgentProfile(profile, currentProfile == null ? RequestType.put : RequestType.post);
```
Comment (mediabounds) [inline Float.TinCan/RemoteLRS.cs:502]: @bbeversdorf IMHO, I think the implementing application is responsible for specifying the `etag`...and it looks like there is already a separate public method (albeit with a questionable name) for using `POST` vs `PUT` (`ForceSaveAgentProfile`).

**Review by mediabounds — CHANGES_REQUESTED** [review]
Two questions in the PR.

File: Float.TinCan/RemoteLRS.cs (hunk 2 above)
Comment (bbeversdorf) [reply Float.TinCan/RemoteLRS.cs:502]: When the SPT was "resaving" the profile and there was already an entry, then we need to POST. This implementation of TinCan _always_ assumes it should be PUT which fails.

Comment (mediabounds) [reply Float.TinCan/RemoteLRS.cs:502]: I'm not sure what the stack trace is leading up to this (maybe this is a change that needs to be made in LocalLRSServer?), but the `ForceSaveAgentProfile` method appears to be the intended way to use `POST`.

File: Float.TinCan/RemoteLRS.cs (hunk 1 above)
Comment (bbeversdorf) [reply Float.TinCan/RemoteLRS.cs:730]: The exception goes into the ether.
(Pseudo stacktrace)
``` 
RemoteLRS.Task<LRSResponse> SaveDocument {
   ...
   MediaTypeHeaderValue.Parse(document.contentType)
}

RemoteLRS.Task<LRSResponse> SaveAgentProfile(AgentProfileDocument profile) {
   ...
   return await SaveAgentProfile(profile, currentProfile == null ? RequestType.put : RequestType.post);
}

LRSProxy.Task<LRSResponse> SaveAgentProfile(AgentProfileDocument profile) {
   .....
   return await targetLRS.SaveAgentProfile(profile).ConfigureAwait(false) 
} // Does not report any where.
````

File: Float.TinCan/RemoteLRS.cs (hunk 2 above)
Comment (bbeversdorf) [reply Float.TinCan/RemoteLRS.cs:502]: Initial agent profiles should be `PUT`. We would still need to know if there was a profile or not.

File: Float.TinCan/RemoteLRS.cs (hunk 1 above)
Comment (bbeversdorf) [reply Float.TinCan/RemoteLRS.cs:730]: The exception does not bubble up to anything. We could try/catch all return statements for `LRSProxy`. 
Currently because there is not an explicit catch we return Status: 204 No Content

File: Float.TinCan/RemoteLRS.cs (hunk 2 above)
Comment (mediabounds) [reply Float.TinCan/RemoteLRS.cs:502]: > We would still need to know if there was a profile or not.

I still think that has to be the responsibility of the caller. The caller might be trying to persist outdated data here; so the caller needs to confirm that it knows the etag of the document it intends to be updating.

`PUT` can still be used for updating an existing document. `POST` is only needed if you want to merge JSON documents.

File: Float.TinCan/RemoteLRS.cs (hunk 1 above)
Comment (mediabounds) [reply Float.TinCan/RemoteLRS.cs:730]: That appears to be an issue in `Float.TinCan.ActivityLibrary`:
https://github.com/gowithfloat/Float.TinCan.ActivityLibrary/blob/5df622093144ab5997b9872334138e4553cfce24/Float.TinCan.ActivityLibrary/TinCanActivityRunner.cs#L280

It is not returning nor awaiting the task for saving the document; since this is done in an event subscriber, task failures will be discarded.

**Comment by bbeversdorf** [comment]
Ok closing in favor of the caller adding ETags and relying on ActivityLibrary to catch the issue
